### ISSUE TITLE
Constants inside functions are not recognised

### ISSUE THREAD

**Issue by svivian** [vimeo/psalm#54]
Test case:

```php
/**
 * @return void
 */
function defineConstant()
{
    define('CONSTANT', 1);
}

defineConstant();

echo CONSTANT;
```
Output:

> ERROR: UndefinedConstant - test-constant.php:13 - Const CONSTANT is not defined
> echo CONSTANT;

Constants are always global, so there is no error in the above script.

**Comment by svivian** [comment]
Also fails in the opposite example:


	define('TEST', 2);

	/**
	 * @return int
	 */
	function useTest()
	{
		return TEST;
	}

Output:

> ERROR: UndefinedConstant - test-constant.php:30 - Const TEST is not defined
> return TEST;

**Comment by muglug** [comment]
Thanks for this – non-trivial to fix the initial version, as it means storing the constants declared by a function. Even then, there'll be a little fudging – I think it should still fail in this scenario:

```php
/**
 * @return void
 */
function defineConstant()
{
    if (rand(0, 1)) {
        define('CONSTANT', 1);
    }
}

defineConstant();

echo CONSTANT;
```

i.e. when the constant is not defined on the function in the main block.

**Comment by muglug** [comment]
In both cases I think the solution is to tend towards false negatives.

**Comment by muglug** [comment]
The latter is now fixed

**Comment by muglug** [comment]
Thanks for detailing these
